crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -972,52 +956,27 @@ impl Object {
     /// Query for a singleton object identified by its type. Note: the object is assumed to be a
     /// singleton (we either find at least one object with this type and then return it, or return
     /// nothing).
-    pub(crate) async fn query_singleton(db: &Db, type_: TypeTag) -> Result<Option<Object>, Error> {
-        use objects::dsl;
-
-        let stored_obj: Option<StoredObject> = db
-            .execute(move |conn| {
-                conn.first(move || {
-                    dsl::objects.filter(
-                        dsl::object_type.eq(type_.to_canonical_string(/* with_prefix */ true)),
-                    )
-                })
-                .optional()
-            })
-            .await
-            .map_err(|e| Error::Internal(format!("Failed to fetch singleton: {e}")))?;
-
-        stored_obj
-            .map(|obj| Object::try_from_stored_object(obj, None))
-            .transpose()
-    }
+    pub(crate) async fn query_singleton(
+        db: &Db,
+        type_: TypeTag,
+        checkpoint_viewed_at: u64,
+    ) -> Result<Option<Object>, Error> {
+        let filter = ObjectFilter {
+            type_: Some(TypeFilter::ByType(type_)),
+            ..Default::default()
+        };
 
-    /// `checkpoint_viewed_at` represents the checkpoint sequence number at which this `Object` was
-    /// constructed in, or `None` if the data was requested at the latest checkpoint. This is
-    /// stored on `Object` so that when viewing that entity's state, it will be as if it was read at
-    /// the same checkpoint.
-    pub(crate) fn try_from_stored_object(
-        stored_object: StoredObject,
-        checkpoint_viewed_at: Option<u64>,
-    ) -> Result<Self, Error> {
-        let address = addr(&stored_object.object_id)?;
-        let native_object = bcs::from_bytes(&stored_object.serialized_object)
-            .map_err(|_| Error::Internal(format!("Failed to deserialize object {address}")))?;
+        let connection = Self::paginate(db, Page::bounded(1), filter, checkpoint_viewed_at).await?;
```

Comment:
Haha this is cheeky, neat stuff

Reply:
Hmm ... all the single object queries are technically just a bounded version of the pagination object query now that I think about it

Reply:
Yes, actually you're pre-empting my big reveal when I implement Object Data Loaders, because they will be implemented by batching up point queries into paginated queries (but with a bound greater than 1, obviously), and we can get rid of the duplication in Object query impls.